cmd/chaos-multiversion-helper/create/create.go

```diff
@@ -0,0 +1,174 @@
+		panic("unreachable")
+	}
+	err = cmd.MarkFlagRequired("to")
+	if err != nil {
+		log.Error(errors.WithStack(err), "fail to mark 'to' as required")
+		panic("unreachable")
+	}
+
+	return cmd
+}
+
+func run(log logr.Logger, from, to string, asStorageVersion bool) error {
+	fileSet := token.NewFileSet()
+
+	oldAPIDirectory := "api/" + from
+	newAPIDirectory := "api/" + to
+	err := os.Mkdir(newAPIDirectory, 0755)
+	if err != nil {
+		return errors.Wrapf(err, "create directory %s", newAPIDirectory)
+	}
+
+	oldFiles, err := ioutil.ReadDir(oldAPIDirectory)
+	if err != nil {
+		return errors.Wrapf(err, "read directory %s", oldAPIDirectory)
+	}
+
+	ctx := createFileContext{
+		oldAPIDirectory,
+		newAPIDirectory,
+		to,
+		asStorageVersion,
+		fileSet,
+		log,
+	}
+	for _, file := range oldFiles {
+		log.Info("copy api", "oldFile", file.Name())
+		ctx.log = log.WithValues("fileName", file.Name())
+		err := createFile(ctx, file)
+		if err != nil {
+			return err
+		}
+	}
+	log.Info("create new api successfully")
+	return nil
+}
+
+type createFileContext struct {
+	oldAPIDirectory string
+	newAPIDirectory string
+	to              string
+
+	asStorageVersion bool
+	fileSet          *token.FileSet
+
+	log logr.Logger
+}
+
+func createFile(ctx createFileContext, file fs.FileInfo) error {
+	// skip files which are not golang source code
+	if !strings.HasSuffix(file.Name(), "go") {
+		return nil
+	}
+
+	oldFilePath := ctx.oldAPIDirectory + "/" + file.Name()
+	fileAst, err := parser.ParseFile(ctx.fileSet, oldFilePath, nil, parser.ParseComments)
+	if err != nil {
+		return errors.Wrapf(err, "parse file %s", oldFilePath)
+	}
+
+	// modify the package name to the new version
+	fileAst.Name.Name = ctx.to
+
+	if ctx.asStorageVersion {
+		// automatically add the storage version annotation
+		cmap := ast.NewCommentMap(ctx.fileSet, fileAst, fileAst.Comments)
+		for node, commentGroups := range cmap {
+			node, ok := node.(*ast.GenDecl)
+			if !ok || node.Tok != token.TYPE {
+				continue
+			}
+
+			for _, commentGroup := range commentGroups {
+				isObjectRoot := false
+				isStorageVersion := false
+
+				for _, comment := range commentGroup.List {
+					if strings.Contains(comment.Text, "+kubebuilder:object:root=true") {
+						isObjectRoot = true
+					}
+					if strings.Contains(comment.Text, "+kubebuilder:storageversion") {
+						isStorageVersion = true
+					}
+				}
+
+				if isObjectRoot && !isStorageVersion {
+					commentGroup.List = append(commentGroup.List, &ast.Comment{
+						Text:  "// +kubebuilder:storageversion",
+						Slash: node.Pos() - 1,
+					})
+				}
+			}
+		}
+
+		// `ast` package is not suitable for removing a comment, so the
+		// traditional string processing tool is preferred :)
+		sedProcess := exec.Command("sed", "-i", "/+kubebuilder:storageversion/d", oldFilePath)
+		err = sedProcess.Run()
+		if err != nil {
+			return errors.Wrapf(err, "remove storage version for %s", oldFilePath)
```

Comment:
And this command does not work on my environment now:

```
2022-06-24T09:53:46.993Z        ERROR   chaos-multiversion-helper       chaos-multiversion-helper/main.go:51    execute command {"error": "remove storage version for api/v1alpha1/awschaos_types.go, msg: sed: couldn't open temporary file api/v1alpha1/sedsI8fNj: Permission denied\n: exit status 4", "errorVerbose": "exit status 4\nremove storage version for api/v1alpha1/awschaos_types.go, msg: sed: couldn't open temporary file api/v1alpha1/sedsI8fNj: Permission denied\n\ngithub.com/chaos-mesh/chaos-mesh/cmd/chaos-multiversion-helper/create.createFile\n\t/Users/siyuchen/Code/chaos-mesh/cmd/chaos-multiversion-helper/create/create.go:164\ngithub.com/chaos-mesh/chaos-mesh/cmd/chaos-multiversion-helper/create.run\n\t/Users/siyuchen/Code/chaos-mesh/cmd/chaos-multiversion-helper/create/create.go:91\ngithub.com/chaos-mesh/chaos-mesh/cmd/chaos-multiversion-helper/create.NewCreateCmd.func1\n\t/Users/siyuchen/Code/chaos-mesh/cmd/chaos-multiversion-helper/create/create.go:43\ngithub.com/spf13/cobra.(*Command).execute\n\t/tmp/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:856\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/tmp/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974\ngithub.com/spf13/cobra.(*Command).Execute\n\t/tmp/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902\nmain.main\n\t/Users/siyuchen/Code/chaos-mesh/cmd/chaos-multiversion-helper/main.go:50\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1263"}
```

The file permission is
```
----------  1 siyuchen  staff     0B Jun 24 17:53 api/v1alpha1/sedsI8fNj
```

I'm not sure if it is related to the new feature on macOS or something, I'll try to take a look, do you have any idea?

Reply:
The file permission is 000, really wired :thinking: 

@STRRL  or @Hexilee  do you know why?

Reply:
@YangKeao It seems there is an issue with VirtioFS (https://github.com/docker/for-mac/issues/6239), it could work after I disable VirtioFS :)

Reply:
> We need to set [the StdErr of the Command Object](https://pkg.go.dev/os/exec#Cmd), if not, it will only print the exit code here https://pkg.go.dev/os/exec#ExitError, which is not friendly for debugging. And exec.Command in other places should set this field too.

I'll use combined output to print both the stdout and stderr if there is an error.

Reply:
Then I think this comment has been fixed? (though I don't know whether the problem in virtiofs will be solved by the upstream)